feat: Introduce evaluation UI and PDF report generation

GrayNekoBean opened the pull request:
## Purpose

Add a frontend page for evaluation and integrate the compare model feature into report generation. 

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

- [ ] Yes
- [x] No

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

- [ ] Yes
- [x] No

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

binxuan39 committed:
feat: add frontend page for evaluation

GrayNekoBean committed:
hide scroll bar

xiatianrui1110 committed:
compare models

xiatianrui1110 committed:
add red teaming comparison feature

xiatianrui1110 committed:
add cli command

xiatianrui1110 committed:
fix graph plotting and red teaming

GrayNekoBean committed:
demo

binxuan39 committed:
feat: Add generate test feature data

- add /generate route
- add generate tab to the first step of the evaluation page

binxuan39 committed:
fix: remove unused package import

xiatianrui1110 committed:
fix red teaming comparison and update readme

xiatianrui1110 committed:
fix red teaming and python test

xiatianrui1110 committed:
fix readme

xiatianrui1110 committed:
fix test snapshots

binxuan39 committed:
fix: front end bug and data structure

- fix error in setting number of questions
- change the request data format of generate to JSON

xiatianrui1110 committed:
in sync with the main branch

xiatianrui1110 committed:
fix python test

xiatianrui1110 committed:
fix python test

GrayNekoBean committed:
Merge remote-tracking branch 'origin/demo' into evaluation-wip-integrate-frontend

GrayNekoBean committed:
minor fix
- fixed error message display issue
- fixed report generation failing bug
- updated .gitignore

binxuan39 committed:
update gitignore

binxuan39 committed:
update gitignore

GrayNekoBean committed:
frontend improvement
- Separated GPT metrics and statistic metrics
- select all by default
- multiple style & layout fixes

GrayNekoBean committed:
Merge branch 'evaluation-wip-integrate-frontend' into dev

GrayNekoBean committed:
Merge branch 'new-compare-models' into dev

GrayNekoBean committed:
fixed return value for test

binxuan39 committed:
fix: red-teaming bugs

- default to use the first model in config when running solely on application
- add type check for common prompt chat target
- add max_turns parameters for red_teaming
- rename the config compare_models fields

GrayNekoBean committed:
update report template in backend

GrayNekoBean committed:
fix: unify 'compared_model' config name

GrayNekoBean committed:
Add model selection for frontend, and other minor improvements
- Add model selection section on evaluation frontend
- Added "compared_models" config in request body
- Added "run red teaming" option
- Minor style improvements

GrayNekoBean committed:
update config for "compared_models"

GrayNekoBean committed:
Enabled red teaming on evaluator frontend
- Enabled run red teaming option in frontend
- Adjusted backend evaluate endpoint
- Added a functin to run both eval
- Aggregated red teaming results into report file
- Minor updates on frontend

GrayNekoBean committed:
make evaluation functions async

GrayNekoBean committed:
enabled evaluation component to call localhost backend, and more asynchronized more functions

GrayNekoBean committed:
use remote backend for default

GrayNekoBean committed:
Merge remote-tracking branch 'origin/main' into dev

GrayNekoBean committed:
add 'include conversation logs' option
- added this option in config to control whether to include conversation section
- added this option in frontend page
- fixed Pass/Fail display bug

GrayNekoBean committed:
removed temp code

GrayNekoBean committed:
fix mypy check

GrayNekoBean committed:
try fix pytest

GrayNekoBean committed:
fix pyrit compatible issue for pytest

binxuan39 committed:
update .gitignore

binxuan39 committed:
Add max turns field for frontend

binxuan39 committed:
fix pytest

binxuan39 committed:
fix: pytest click error

GrayNekoBean committed:
moved evaluation directory into app/backend

binxuan39 committed:
add test case for evaluation

GrayNekoBean committed:
Update path to evaluation dir

GrayNekoBean committed:
Merge branch 'dev' of github.com:ucl-contoso-chat/ucl-openai-search into dev

GrayNekoBean committed:
update path in launch.json

GrayNekoBean committed:
fetch BACKEND_URL from env var

GrayNekoBean committed:
remove .env.sample

GrayNekoBean committed:
revert mis-overrided code

binxuan39 committed:
test: add plotting graph test, remove import ignore

phoevos committed:
Merge branch 'main' into dev

phoevos reviewed (changes requested):
I still haven't had the chance to properly look through the PR, but until I do I would really like to see issues I've mentioned before fixed. The main one is discrepancies between the PR code and `main`. If something doesn't need to change, it shouldn't. Please take a close look at the code (especially in cli.py, and evaluate.py) and the evaluation README.

There are minor stuff like the mistake in the `--num-questions` CLI args in `.vsode/launch.json`, as well as things I've bitched about in literally every PR before, like:
* Every function/method should have a docstring
* Docstrings should follow the [Google styleguide conventions](https://google.github.io/styleguide/pyguide.html#381-docstrings) like starting with a capital, ending with a period, no newline between them and the code, as well as general conventions like being formulated in the imperative
* There should be no commented-out code left behind
* The CI should be passing. Just merged some changes from main, fingers crossed it will improve the situation

Also, mentioning again that we're not using the `load_env` scripts and that the format of `summary.json` shouldn't change.

Apart from that, I noticed during the demo that the header of our page ("Chat" "Ask a Question", "Evaluate") is not evenly spaced. Although I won't go into too much depth on the frontend side, I do believe that this one should be fixed.

Finally, did you handle properly generating the environment needed for the evaluation to run? Are you running `create_eval_dotenv`? Are you fetching the values in a different way? How are they being passed to the evaluation script?

Regardless of all of the above, have you tested a fresh deployment?

Given that the PR was opened 3 days before the submission deadline it's unlikely that we'll be able to do a proper review. But I do expect that all superficial issues are resolved. Also just to be crystal clear: there is extreme code duplication for the "evaluation by request" feature and code has been moved around to facilitate just that (e.g. constants being moved into the evaluate module although they clearly don't belong there, a new function in evaluate running red teaming stuff- not the place to put it most likely- let alone repeating the whole logic and configuration parsing done in the existing functions meant to run the evaluation). This is ok if writing code that needs to be demoed in 2 hours, but not quite PR material. I kindly ask that we push to improve this PR as much as possible before the deadline.

GrayNekoBean commented:
> there is extreme code duplication for the "evaluation by request" feature and code has been moved around to facilitate just that (e.g. constants being moved into the evaluate module although they clearly don't belong there, a new function in evaluate running red teaming stuff- not the place to put it most likely- let alone repeating the whole logic and configuration parsing done in the existing functions meant to run the evaluation).

Thanks, I also just noticed that those two functions' behaviour was more similar than I thought like an hour ago and already removed one of them. (more specifically, combined them) Removing that should also rise the coverage.

Me and binxuan will get these fixed asap, but we'll also keep the changes where we see necessary.

GrayNekoBean committed:
combined two run_evaluation functions

GrayNekoBean committed:
remove commented code

GrayNekoBean committed:
fix import for ruff check

binxuan39 commented:
> I still haven't had the chance to properly look through the PR, but until I do I would really like to see issues I've mentioned before fixed. The main one is discrepancies between the PR code and `main`. If something doesn't need to change, it shouldn't. Please take a close look at the code (especially in cli.py, and evaluate.py) and the evaluation README.
> 
> There are minor stuff like the mistake in the `--num-questions` CLI args in `.vsode/launch.json`, as well as things I've bitched about in literally every PR before, like:
> 
> * Every function/method should have a docstring
> * Docstrings should follow the [Google styleguide conventions](https://google.github.io/styleguide/pyguide.html#381-docstrings) like starting with a capital, ending with a period, no newline between them and the code, as well as general conventions like being formulated in the imperative
> * There should be no commented-out code left behind
> * The CI should be passing. Just merged some changes from main, fingers crossed it will improve the situation
> 
> Also, mentioning again that we're not using the `load_env` scripts and that the format of `summary.json` shouldn't change.
> 
> Apart from that, I noticed during the demo that the header of our page ("Chat" "Ask a Question", "Evaluate") is not evenly spaced. Although I won't go into too much depth on the frontend side, I do believe that this one should be fixed.
> 
> Finally, did you handle properly generating the environment needed for the evaluation to run? Are you running `create_eval_dotenv`? Are you fetching the values in a different way? How are they being passed to the evaluation script?
> 
> Regardless of all of the above, have you tested a fresh deployment?
> 
> Given that the PR was opened 3 days before the submission deadline it's unlikely that we'll be able to do a proper review. But I do expect that all superficial issues are resolved. Also just to be crystal clear: there is extreme code duplication for the "evaluation by request" feature and code has been moved around to facilitate just that (e.g. constants being moved into the evaluate module although they clearly don't belong there, a new function in evaluate running red teaming stuff- not the place to put it most likely- let alone repeating the whole logic and configuration parsing done in the existing functions meant to run the evaluation). This is ok if writing code that needs to be demoed in 2 hours, but not quite PR material. I kindly ask that we push to improve this PR as much as possible before the deadline.

binxuan39 commented:
> There are minor stuff like the mistake in the --num-questions CLI args in .vsode/launch.json, as well as things I've bitched about in literally every PR before, like:

The correct one should be connected like we using now as the Typer automatically converts the underline to the hyphen.

binxuan39 committed:
relocate the dir variables

binxuan39 committed:
remove commented code

binxuan39 committed:
fix start.sh/ps1

binxuan39 committed:
fix: frontend heder tab spacing

binxuan39 committed:
edit start.sh/ps1

binxuan39 committed:
fix: BACKEND_URI typo

phoevos committed:
Merge branch 'main' into dev

GrayNekoBean committed:
minor fixes for deployment

GrayNekoBean committed:
set BACKEND_URL env var

GrayNekoBean committed:
fix requirements-dev.txt path

phoevos committed:
Merge branch 'main' into dev

phoevos commented on app/backend/app.py:
```diff
@@ -339,6 +360,97 @@ async def speech():
+        config["testdata_path"] = "input/input_temp.jsonl"
+        config["results_dir"] = "results"
+
+        temp_config_path = Path("config_temp.json")
+
+        save_config(config, "evaluation" / temp_config_path)
```
Looks like you would be overwriting the same config file if more than one requests come in. On top of that, it doesn't look like you ever delete it.

GrayNekoBean replied:
It looks like the temp file will be loaded instantly after been written so I guess it doesn't really affect the running result in normal situations (these lines look a bit dumb though). They should be some sort of legacy code, I'll handle it.

phoevos committed:
fix: Update project dependencies

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! fix: Update project dependencies

Shame that we have to revert to something like that, but this is far
from our top priority right now. We will revisit this in the future.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos commented on .github/workflows/model-evaluation.yml:
We need to add the report generation functionality here (as well as the corresponding artifact.)

phoevos committed:
fix: Standardise blank lines in plotting.py

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

GrayNekoBean committed:
removed config_temp.json generation

phoevos committed:
fix: Update async model fetching

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! removed config_temp.json generation

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos commented on .gitignore:
```diff
@@ -153,6 +146,17 @@ npm-debug.log*
 data/**/*.md5
 
+# Evaluation
+app/backend/evaluation/report
```
Is anything ever generated in there? AFAICT, unless we explicitly specify the path, this should go under the timestamped experiment directories.

GrayNekoBean replied:
This is only used in some local machine tests, but its okay to remove that

phoevos committed:
fix: Update .gitignore

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Drop integration mistake in evaluation README

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Add missing requirement to evaluation .in file

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

GrayNekoBean committed:
fix: update model-evaluation CI command for proper report generation

phoevos commented:
I'm not sure about the whole process of adding the red teaming in the evaluation bit. The combined process was always supposed to be a separate command that runs both and generates the report. I was expecting a new function that calls what we already had there in the evaluation and red teaming modules instead of patching up the evaluate part to do some extra things it wasn't intended to do. I'll try to do this refactoring myself, cause this keeps going down the wrong path.

PS: Deprecating the red-teaming command and only exposing a single evaluation that takes care of everything is absolutely an option, but it was never discussed.

phoevos committed:
fix: Run codespell for spelling issues

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

GrayNekoBean commented:
> I'm not sure about the whole process of adding the red teaming in the evaluation bit. The combined process was always supposed to be a separate command that runs both and generates the report. I was expecting a new function that calls what we already had there in the evaluation and red teaming modules instead of patching up the evaluate part to do some extra things it wasn't intended to do. I'll try to do this refactoring myself, cause this keeps going down the wrong path.
> 
> PS: Deprecating the red-teaming command and only exposing a single evaluation that takes care of everything is absolutely an option, but it was never discussed.

I totally get what you said, ideally I should be making a new function that invokes both parts and I was trying to do that in the beginning.  But that means I need to move report generation lines to an upper layer of the function call, and I was afraid that could make our code look more nested. I could do the refactor if the services are all good, but now I don't dare to make huge changes without testing it.

phoevos committed:
fix: Major restructuring

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

binxuan39 committed:
fix: type error

nextedoff committed:
fix: mypy fix

Signed-off-by: nextedoff <rajmukx22@gmail.com>

phoevos committed:
Merge branch 'main' into dev
